### DISCUSSION TITLE
spaCy multiprocessing ('n_process') makes Docker image exit

### DISCUSSION THREAD

**Discussion by dave-espinosa** [explosion/spaCy#10087] [Help: Coding & Implementations]
Hello everyone,

Long story short: I need to apply some NER on several thousands of documents. I have decided to go with the "en_core_web_lg" model. I want to increase the NER processing speed as much as possible, so the [speed FAQ](https://github.com/explosion/spaCy/discussions/8402) has been already considered. Two of the main suggestions there, require the developer to use `nlp.pipe` and fine-tune the `n_process` argument in it, which I have already done, to obtain something like this:

```
import spacy
import os
nlp = spacy.load("en_core_web_lg")
raw_skill_list = [list(doc.ents) for doc in nlp.pipe(text_list, batch_size=17, n_process=os.cpu_count()-1)]
# NOTES:
# 1) `text_list` contains a long, long list of documents
# 2) The original code includes the disabling of some pipeline components, I have done that, but for the sake of this issue, it is not important
```

The results look fine so far, at least in the tests ran in my computer. To speed things up even more, and since my company works with GCP, I was planing to create a Docker image with the previous code in it, to then deploy it in several endpoints. The goal is of course split the 'text_list' in 'n' chunks, where `n` = number of endpoints. The problem that I have right now is that when `n_process` = 1, the Docker image manages to normally process all the texts, batch after batch (in the following toy example, 2 batches of 100 texts were used):

```
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
INFO:     172.17.0.1:35560 - "GET / HTTP/1.1" 200 OK
INFO:     172.17.0.1:35560 - "GET /docs HTTP/1.1" 200 OK
INFO:     172.17.0.1:35560 - "GET /openapi.json HTTP/1.1" 200 OK
Skills extracted from 100 texts, in 0.28 seconds
Skills cleaned from 100 lists, in 0.00 seconds
Skills refined from 100 lists, in 1.29 seconds
INFO:     172.17.0.1:35564 - "POST /eng/plain/ HTTP/1.1" 200 OK
Skills extracted from 100 texts, in 0.28 seconds
Skills cleaned from 100 lists, in 0.00 seconds
Skills refined from 100 lists, in 1.29 seconds
INFO:     172.17.0.1:35564 - "POST /eng/plain/ HTTP/1.1" 200 OK
```

BUT if `n_process` > 1, then the image just processes the first batch, and then exits:

```
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
Skills extracted from 100 texts, in 0.21 seconds
Skills cleaned from 100 lists, in 0.00 seconds
Skills refined from 100 lists, in 1.32 seconds
INFO:     172.17.0.1:35792 - "POST /eng/plain/ HTTP/1.1" 200 OK
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
```

After some research in Google, even when "one process per Docker image" is usually recommended, [there are ways](https://stackoverflow.com/questions/37871540/how-many-cpus-does-a-docker-container-use) to achieve multiprocessing. However and quite honestly I don't know how to adapt those suggestions for spaCy's `nlp.pipe`.

Have you had any experience implementing something similar to this?

PS.: I discarded [this recommendation](https://github.com/explosion/spaCy/issues/4617#issuecomment-557801099) because my company's GCP architecture is better oriented to deal with Docker images.

Thank you.

**Comment by pmbaumgartner** [comment]
How many workers are you launching `uvicorn` with? I'm guessing there's a conflict there with the workers & `n_process`.

Is it possible to share the `Dockerfile` you're using?

**Reply by dave-espinosa** [reply]
Hello @pmbaumgartner ,

> How many workers are you launching `uvicorn` with? I'm guessing there's a conflict there with the workers & `n_process`.

- In the local tests I run `uvicorn src.app:app --reload`. I think this does not matter, since the [documentation](https://www.uvicorn.org/deployment/) says `workers` are "_Not valid with --reload_"...
- In my original Dockerfile, I first included `uvicorn src.app:app --host 0.0.0.0 --port 8080`. In the [documentation](https://www.uvicorn.org/deployment/) says `workers` "_Defaults to the $WEB_CONCURRENCY environment variable if  available, or 1_". I do not have such `$WEB_CONCURRENCY` environment variable in my Dockerfile, for which I think it was set to 1. However, I decided to build another image with `uvicorn src.app:app --host 0.0.0.0 --port 8080 --workers 8` in it, and **now the image does not exit anymore** 🙂 ...

... BUT results are far from satisfactory ☹, because for some reason **I am getting results that take way longer to process than before**. For instance, processing a **single batch of 100 texts**, now **it takes 77.16 seconds**, when **the original setting** managed to process the same amount of information, in **0.28 seconds** (see original post):

```
INFO:     Started server process [12]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     172.17.0.1:33038 - "GET / HTTP/1.1" 200 OK
INFO:     172.17.0.1:33038 - "GET /favicon.ico HTTP/1.1" 404 Not Found
INFO:     172.17.0.1:33038 - "GET /docs HTTP/1.1" 200 OK
INFO:     172.17.0.1:33038 - "GET /openapi.json HTTP/1.1" 200 OK
Skills extracted from 100 texts, in 77.16 seconds
Skills cleaned from 100 lists, in 0.03 seconds
Skills refined from 100 lists, in 1.56 seconds
INFO:     172.17.0.1:33046 - "POST /eng/plain/ HTTP/1.1" 200 OK
Skills extracted from 100 texts, in 64.17 seconds
Skills cleaned from 100 lists, in 0.00 seconds
Skills refined from 100 lists, in 1.79 seconds
INFO:     172.17.0.1:33050 - "POST /eng/plain/ HTTP/1.1" 200 OK
```

The previous results were run from my computer. And just in case you were wondering, I am currently working with 7 out of my 8 CPU cores (`n_process` = 7), so I _intuitively_ set `workers` = 8, in my experiments before.

> Is it possible to share the `Dockerfile` you're using?

Yes, I basically used the following [template](https://github.com/mrn18d/example-nlp-project/blob/main/Dockerfile), with the only difference that I am using `python:3.9-slim` and `POETRY_VERSION=1.1.12` (and of course, `nlp.pipe` to process the text batches, as explained in my original post). BTW, you can use that project as reference, since my current implementation is very similar.

Please let me know your thoughts, and best regards.

Thank you.

**Reply by pmbaumgartner** [reply]
@dave-espinosa Thanks for all the detail, that will be really helpful in replicating and debugging the issue. I'm guessing it'll take me some time to try things out but I'm going to take a look at this soon and will get back to you.

**Reply by pmbaumgartner** [reply]
@dave-espinosa Alright, I've had a bit of time to look at the speed / processing issue but haven't dug into the docker issue yet.

On the speed point, I think the batch size you have set up is the opposite direction we want to go: we want a larger batch size. There's a bit on this in the docs [here](https://spacy.io/usage/processing-pipelines#multiprocessing) why this is the case.

> The results look fine so far, at least in the tests ran in my computer. To speed things up even more, and since my company works with GCP, I was planing to create a Docker image with the previous code in it, to then deploy it in several endpoints. 

> After some research in Google, even when "one process per Docker image" is usually recommended, there are ways to achieve multiprocessing. However and quite honestly I don't know how to adapt those suggestions for spaCy's nlp.pipe.

I'll admit I'm not _that_ experienced in this area, but when I've seen distributed systems like this implemented you actually do want one process per image and to rely on your orchestration to create multiple containers with the same (single process) image. So rather than have one docker image with multiprocessing, you'd have one image but several containers running that image, and then you'd do the distributed part at that layer rather than within the image at the `nlp.pipe` layer. In this case you'd be pre-partitioning or batching your data, possibly using a tool like [ray](https://www.ray.io/) or [dask](https://dask.org/), then passing it to a worker/container.

In the end, your implementation per-worker / per-container will look closer to [this](https://github.com/explosion/spaCy/issues/5111#issuecomment-671383361). 

I'm still planning on taking a look at the Docker issue to see if there's anything interesting there, but, if you create a single-process image using my suggestion above and don't use `n_process`, that also solves that problem for us too! :)

**Reply by pmbaumgartner** [reply]
For replicating the docker issue I need a bit more information. I'd need a a `src` folder with the contents of your app as well as your `pyproject.toml` and `poetry.lock` file, since I'm not sure what versions of packages you're using.

**Reply by dave-espinosa** [reply]
Hello @pmbaumgartner ,

Regarding some of your comments:

> On the speed point, I think the batch size you have set up is the opposite direction we want to go: we want a larger batch size. There's a bit on this in the docs [here](https://spacy.io/usage/processing-pipelines#multiprocessing) why this is the case.

Allow me to update on that: trying to find the best values for _our_ use case, we first ran our experiments with _real-life samples_, that have ~5000 characters or so (roughly one page of text) each. And the results seem... _weird_. See for instance some quick results I ran on a batch, consisting of 1000 texts:

```
Working on 'batch_size'=10... Done! (Elapsed time: 95.58[s])
Working on 'batch_size'=20... Done! (Elapsed time: 90.01[s])
Working on 'batch_size'=30... Done! (Elapsed time: 89.22[s])
Working on 'batch_size'=40... Done! (Elapsed time: 91.06[s])
Working on 'batch_size'=50... Done! (Elapsed time: 91.94[s])
Working on 'batch_size'=100... Done! (Elapsed time: 91.96[s])
Working on 'batch_size'=200... Done! (Elapsed time: 93.44[s])
Working on 'batch_size'=300... Done! (Elapsed time: 94.17[s])
Working on 'batch_size'=400... Done! (Elapsed time: 95.15[s])
Working on 'batch_size'=500... Done! (Elapsed time: 95.00[s])
```

Notice how the quickest processing time, is obtained when the `batch_size` is "30", even when larger `batch_size` values are available (seems "reasonable" anyways, because our texts are somehow lengthy). Notice as well how a `batch_size` of "10" has the same processing speed as one with "400" as `batch_size`. Finally, the value "17" was the _absolute best_ in our experiments (NOT portrayed in this demo example), and so that is why it was included that way. Thank you anyways, for the [docs](https://spacy.io/usage/processing-pipelines#multiprocessing) reminder ;).

**NOTE: For the sake of your experiments, the only thing that matters, is that test texts, must have a lenght of roughly one page**.

> For replicating the docker issue I need a bit more information. I'd need a a `src` folder with the contents of your app as well as your `pyproject.toml` and `poetry.lock` file, since I'm not sure what versions of packages you're using.

Gladly, [here](https://github.com/dave-espinosa/ner_demo) you go.

Please let me know if you find anything new. For the time being, and since we cannot wait any longer, we decided to go on with this suggestion, as a **containment** solution (read: we still need to check if we can squeeze spaCy's performance even more):

> So rather than have one docker image with multiprocessing, you'd have one image but several containers running that image, and then you'd do the distributed part at that layer rather than within the image at the nlp.pipe layer. In this case you'd be pre-partitioning or batching your data, possibly using a tool like ray or dask, then passing it to a worker/container.

Thank you very much, and have a nice week.

**Reply by pmbaumgartner** [reply]
Thanks again for all the detail. I'm glad to hear you've got a solution that will help you move forward -- it's always good to get things working and then optimize.

I actually took this issue and ran some experiments with `n_process` and `batch_size` myself, and can confirm some of what you've seen. In particular, `batch_size` didn't have a have a huge impact for me.

I also experienced really _slow_ performance if `n_process!=1`, and can boil it down to working on macOS. Are you working on a Mac for your testing? Due to how multiprocessing in python>3.8 [works on macOS](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods) it can actually make things slower (it has to start up a new python interpreter process) to use `n_process>1`. In my tests this also extended to running the benchmarks in a python docker container. This is particularly true with the `en_core_web_lg` model as there's a significant startup cost for each new process to inherit that model. I was testing on a not-that-powerful Macbook Air, so that could also be the source of some slowness here if its under-resourced. On that point, be sure to check how much CPU and RAM you're allocating to docker through the preferences there -- I've been bit by the defaults with that before.

I additionally ran the experiments on a dedicated linux box and can confirm that `n_process>1` does provide a speedup in that situation. If you can test your solution on a dedicated linux machine as well, based on my benchmarks I got at least a `1.5x` speedup going from `1` to `2` for `n_process`, then incremental improvements beyond there. 

Those incremental improvements from my benchmark make me think the optimal solution for your system might be to fine-tune the resources per-container and set `n_process=2` for each of them (rather than one big box with `n_process>>1`). But again, this isn't my area of expertise so there may be something even better.

**Reply by dave-espinosa** [reply]
Hello @pmbaumgartner ,

Regarding some of your queries:

> Are you working on a Mac for your testing?

No, here are the specs of the machine I am currently working with:

- Processor: Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz   2.30 GHz (8 CPU cores)
- Installed RAM:	32.0 GB (31.9 GB utilizable)
- System type:  64 bits operating system, x64 processor
- Operating system: Windows 10 Pro
- Version: 21H2
- OS compilation: 19044.1466
- Experience: Windows Feature Experience Pack 120.2212.3920.0

Unfortunately, since I am working remotely, I have only that machine "ready to work", however I am doing my research, to know how to use a GCP VM to deal with Docker as well as GitHub repos, and check if I can use its power to do my tests (as far as I know, those machines use Ubuntu as OS). This might take time from my side, however some colleague(s) will be working next week on this issue, so expect _new attendants to this thread_ 😁.

> Those incremental improvements from my benchmark make me think the optimal solution for your system might be to fine-tune the resources per-container and set `n_process=2` for each of them (rather than one big box with `n_process>>1`). But again, this isn't my area of expertise so there may be something even better.

My colleague(s) who will be working next week on this research, already has(have) the link to this thread, so (t)he(y) will probably include his(their) results and see if we can notice any improvement from our side. And hopefully, we reach some new discovery on this regard. Granted, they will be running those tests from a linux-based OS, as per your suggestion.

Best regards, and hope to let you know what happened soon.

Thank you very much,

**Reply by pmbaumgartner** [reply]
@dave-espinosa The more the merrier! Looking forward to figuring things out.

One additional note: I neglected to mention my implication about [multiprocessing](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods) being slow on macOS also applies to Windows, so I'm not yet convinced we can rule that out as the cause of the local slowdown.

**Reply by adestevez** [reply]
Hi @dave-espinosa  @pmbaumgartner 

I have performed the experiments as mentioned in the corresponding thread.  However, prior to that neatly on an "Ubuntu 20.04.3 LTS OS, 64 bits, RAM 7,6 GiB, Intel® Core™ i7-8565U CPU @ 1.80GHz × 8, Mesa Intel® UHD Graphics 620 (WHL GT2),  ", the results were as follows:

- Skills extracted from 10 texts, in 0.35 seconds
- Skills extracted from 100 texts, in 3.25 seconds
- Skills extracted from 1000 texts, in 33.98 seconds

The results obtained according to the comments posted in the thread, for **4 cores 0.59 seconds**, while for **8 cores a time of 0.93 seconds.**

It is important to mention that the workers were not used with the experiments since they initially worsened the experiments.

**Reply by dave-espinosa** [reply]
Hello @adestevez ,

Thank you very much for your experiments.

@pmbaumgartner , in general terms, the following insights can be extracted from Alexis' experiments:

- **spaCy:**
  - It seems that a linux-based OS does provide a boost in speed on the NER pipeline. But if this is the case, shouldn't some warning be present in spaCy's documentation about this, so the developer is aware that there _could_ be time differences in speed tests, if the OS is changed (as we have discovered in this thread)?
- **Others:**
  - Not including the `workers` parameter (uvicorn) provides _unconclusive / unstable_ behaviours when used with spaCy pipeline, if `n_process>1`. In Windows for instance, it caused to process only the first text batch and then exist the Docker image; in Ubuntu it causes to make pipeline even slower than not using it.

I fear we are reaching the limits of spaCy's performance, but the explanation as to _why_ is this happening, is getting _darker & spookier_ 👻😨... Just joking here 😉, sorry!

If you have some other clue, or a different result from some experiment on your side, please let us know.

Thank you!

**Reply by pmbaumgartner** [reply]
Realized I had replied at the top level and not in thread, apologies!

https://github.com/explosion/spaCy/discussions/10087#discussioncomment-2090269

**Comment by pmbaumgartner** [comment]
@dave-espinosa That all makes sense to me, thanks for summarizing.

We do have a note in the docs about the behavior, but it's not explicitly called out. We do include a link that I referenced throughout to explain what's going on.

> Not including the workers parameter (uvicorn) provides unconclusive / unstable behaviours when used with spaCy pipeline, if n_process>1. In Windows for instance, it caused to process only the first text batch and then exist the Docker image; in Ubuntu it causes to make pipeline even slower than not using it.

The issue here is that `workers` and `n_process` both use multiprocessing. For example, If you have `workers=4`, that's 4 processes that you're running. Inside each of those workers, if you've set `n_process` to anything other than `1`, that worker is also going to start additional subprocesses. Quickly you'll see this adds up to significant overhead as you now have `uvicorn` managing `workers` managing `subprocesses`.

In addition, it's easy to run out of memory in these scenarios, so you might want to look into that. The docker defaults are pretty low for memory use and typically when RAM is the issue you'll see the exit behavior like you have.

**Reply by dave-espinosa** [reply]
Thank you very much for your quick response @pmbaumgartner ,

Understood. I think we have got to a _dead end_ here. There could be some space for improvement, but that would imply modifying current's spaCy source code to get full advantage of `multiprocessing` Python library (we will try to include those experiments, but they won't be a priority as of now). To wrap up, we will try to stick to the safe side, i.e.:

- Keep `n_process=1` and tune speed only using `batch_size`.
- Try to use a linux-based OS for the endpoints.
- Split in as much endpoints as possible, to increase the NER processing speed even more.

Let me know if some other suggestion is being left-out in my checklist.

Thank you very much!
